src/System.Management.Automation/engine/ExecutionContext.cs

```diff
@@ -1429,6 +1439,88 @@ internal static Assembly LoadAssembly(string name, string filename, out Exceptio
+
+            // Proceed loading if assembly is found in GAC.
+            if(!String.IsNullOrEmpty(gacFileName))
+            {
+                try
+                {
+                    loadedAssembly = ClrFacade.LoadFrom(gacFileName);
+
+                    // Get exported types to force dependency loading.
+                    // If a dependency is missing loadedAssembly.ExportedTypes will throw FileNotFoundException.
+                    var exportedTypes = loadedAssembly.ExportedTypes;
+                }
+                catch(FileNotFoundException fileNotFound)
+                {
+                    // Missing dependency assembly full name is in the inner exception.
+                    string dependencyFileName = ((FileNotFoundException)fileNotFound.InnerException)?.FileName;
+                    // Ignore the return value of dependencies so that the exported types of
+                    // the cmdlet assembly are correct.
+                    LoadAssemblyFromGac(dependencyFileName);
+                }
+            }
+
+            return loadedAssembly;
+        }
+
+        private static string GetAssemblyFixedName(string name)
+        {
+            return name.EndsWith(".dll", StringComparison.OrdinalIgnoreCase)
+                        ? Path.GetFileNameWithoutExtension(name)
+                        : name;
+        }
+
+        // Search the GAC for the assembly.
+        // First try MSIL and then try the GAC_<PROCESS_ARCH>
+        private static string GetValidGacPath(string assemblyName)
+        {
+            // Get name of assembly without '.dll'. This is the folder name in GAC.
+            // In some cases '.dll' might not be present at the end.
+            string fixedName = GetAssemblyFixedName(assemblyName);
+
+            string winDirPath = Environment.GetEnvironmentVariable("WINDIR");
+            System.Text.StringBuilder gacPath = new System.Text.StringBuilder();
+            gacPath.AppendFormat("{0}\\Microsoft.Net\\assembly\\GAC_MSIL\\{1}", winDirPath, fixedName);
+
+            if(!Utils.NativeDirectoryExists(gacPath.ToString()))
+            {
+                gacPath.Clear();
+
+                if(Environment.Is64BitProcess)
+                {
+                    gacPath.AppendFormat("{0}\\Microsoft.Net\\assembly\\GAC_64\\{1}", winDirPath, fixedName);
+                }
+                else
+                {
+                    gacPath.AppendFormat("{0}\\Microsoft.Net\\assembly\\GAC_32\\{1}", winDirPath, fixedName);
+                }
+
+                if(!Utils.NativeDirectoryExists(gacPath.ToString()))
+                {
+                    return null;
+                }
+            }
+
+            var fileToSearch = fixedName + ".dll";
+            var fileList = Directory.GetFiles(gacPath.ToString(), fileToSearch, System.IO.SearchOption.AllDirectories);
+
+            if(fileList.Length > 0)
+            {
+                // GAC only has 1 file in the folder, so we can pick index 0.
```

Comment:
Several versions of same assembly?
What if a binary module references v2 of an assembly, when GAC has v1 and v2 of that assembly installed?